feat(csp): add hosted CSP header (#13926)
withastro/astro: .changeset/rare-ducks-sip.md

---
'astro': patch
---

Adds a new **experimental** Astro Adapter Feature called `_experimentalStatiHeaders`. 

> When enabled, and the `experimental.csp` feature is enabled, Astro **won't** serve the CSP `<meta http-equiv="content-security-policy">` element in static pages;
> instead the value of the header is served inside a map that can be retrieved from the hook `astro:build:generated`.
> 
> The name of the new field is called `_experimentalCspMapping` and it contains a map of `Map<IntegrationResolvedRoute, string>` where
> the `string` type is the value of the heder for the current route e.g. `"script-src: 'self' sha256-yadayada; style-src: 'self' sha256-yadayada"

Reviewer: ``````suggestion
For adapters that enable support for this feature, Astro will not serve the CSP `<meta http-equiv="content-security-policy">` element in static pages.

Instead, Astro will serve the value of the header inside a map that can be retrieved from the hook `astro:build:generated`. Adapters can read this mapping and use their hosting headers capabilities to create a configuration file.

A new field called `_experimentalCspMapping` will contain a map of `Map<IntegrationResolvedRoute, string>` where the `string` type is the value of the header for the current route (e.g. `"script-src: 'self' sha256-abc123; style-src: 'self' sha256-abc123").

To enable support for this experimental Astro Adapter Feature, add it to your `supportedAstroFeatures` in your adapter config:

```diff
// my-adapter.mjs
 export default function createIntegration() {
  return {
    name: '@example/my-adapter',
    hooks: {
      'astro:config:done': ({ setAdapter }) => {
        setAdapter({
          name: '@example/my-adapter',
          serverEntrypoint: '@example/my-adapter/server.js',
          supportedAstroFeatures: {
            _experimentalStaticHeaders: 'experimental'
          }
        });
      },
    },
  };
}
```

You must also configure `experimental.csp` in your Astro config:

```js
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  experimental: {
    csp: true,
  },
});
``````
Maybe something like this? I don't know whether the feature itself is named with the underscore? and the level of support is `experimental`? Please make that accurate, but we should show how to configure the adapter here, I think!

Author: I will update your suggestion, and the code should be updated.

---
'astro': patch
---

Adds a new Astro Adapter Feature called `experimentalStaticHeaders` to allow your adapter to control serving CSP in static pages.

For adapters that enable support for this feature, Astro will not serve the CSP `<meta http-equiv="content-security-policy">` element in static pages.

Instead, Astro will serve the value of the header inside a map that can be retrieved from the hook `astro:build:generated`. Adapters can read this mapping and use their hosting headers capabilities to create a configuration file.

> The name of the new field is called `experimentalRouteToHeaders` and it contains a map of `Map<IntegrationResolvedRoute, Headers>` where
> the `Headers` contain the headers emitted during the rendering of the static pages.
> 
> A new field called `experimentalRouteToHeaders` will contain a map of `Map<IntegrationResolvedRoute, Headers>` where the `Headers` type contains the headers emitted by the rendered static route. 

Reviewer: ```suggestion
A new field called `experimentalRouteToHeaders` will contain a map of `Map<IntegrationResolvedRoute, Headers>` where the `Headers` type contains the headers emitted by the rendered static route. 
```
Looks like an old and new version of a sentence here. If correct, the second one is the better sentence!

To enable support for this experimental Astro Adapter Feature, add it to your `supportedAstroFeatures` in your adapter config:

```js
// my-adapter.mjs
export default function createIntegration() {
  return {
    name: '@example/my-adapter',
    hooks: {
      'astro:config:done': ({ setAdapter }) => {
        setAdapter({
          name: '@example/my-adapter',
          serverEntrypoint: '@example/my-adapter/server.js',
          adapterFeatures: {

>             experimentalRouteToHeaders: true

Reviewer: I think so?

          }
        });
      },
    },
  };
}
```

To use this adapter feature, users must also have the `experimental.csp` feature flag set in their Astro config:

```js
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  experimental: {
    csp: true,
  },
});

> ```

Reviewer: ``````suggestion
```

See the [Adapter API docs](https://docs.astro.build/en/reference/adapter-reference/#adapter-features) for more information about providing adapter features.
``````
Maybe something like this?
